[쇼핑 주문 3, 4단계] 크림 미션 제출합니다.

ijh1298 opened the pull request:
두루, 안녕하세요! 소중한 시간 내어 코드 리뷰해주셔서 감사합니다.

여전히 이전 피드백을 다 반영하지 못하고, 버그, 테스트 코드 등 빈약한 부분이 많습니다.
리팩터링에 매진하겠습니다💪

# 셀프 체크리스트

 <!--
 모든 항목을 점검한 후 리뷰를 요청하면, 보다 의미 있는 피드백을 받을 수 있습니다.
 -->

- [ ] 프로그램이 정상적으로 작동하는가?
- [ ] 모든 테스트가 통과하는가?
- [ ] 이전에 받은 피드백을 모두 반영하였는가?
- [x] 코딩 스타일 가이드를 준수하였는가?
    - [x] IDE 코드 자동 정렬을 적용하였는가?
    - [x] 린트 검사를 통과하였는가?
- [x] 프로그래밍 요구 사항을 준수하였는가?
- [x] `README.md`에 기능 목록을 정리하고 명확히 기술하였는가?

# 어떤 부분에 집중하여 리뷰해야 할까요?

 <!--
 학습 목표에 맞게 구현하였는지 점검하고, 피드백을 받을 준비가 되었는지 확인하세요.
 정답을 묻기보다, 자신의 생각과 의사 결정 과정을 공유하며 피드백을 통해 논의하는 것이 좋습니다.
 -->
 
- Repository, DataSource 분리를 처음 진행해봤는데 이에 대해 미숙한 점이 있을 듯합니다.
- 콜백 함수들을 코루틴으로 리팩터링했는데 이에 대해 미숙한 점이 있을 듯합니다.

## 코드 리뷰 커뮤니케이션

**아래는 이전 미션 피드백 답변입니다!**

> ListAdapter + DiffUtils의 적용 아주 좋습니다~
이전 RecyclerView.Adapter와 비교했을 때 어떤 차이점이 있었을까요?!
크림의 자유로운 의견을 남겨주세요ㅎㅎ

ListAdapter + DiffUtils를 사용하니 삭제나 변경된 데이터에 대해 개발자가 notify 해주지 않아도, DiffUtil이 알고리즘을 이용해 알아서 변화를 알려준다는 점이 사용하기 무척 편했습니다. notify range를 지정해줄 때 index 하나 차이로도 OutOfBound 예외가 발생하는 경험을 자주 했는데, ListAdapter로 이전하고 나서 그런 면을 관리할 필요가 줄어들어서 안정적임을 느꼈어요.

다만, DiffUtil이 잘 짜여진 알고리즘이더라도 일부 상황에서는(e.g. 하나의 아이템이 변경된 상황) 직접 변경된 index를 notify 해주는 것이 성능적으로 더 좋을 거라 생각합니다!

> 이 부분은 INCREASE, DECREASE 이렇게 두개의 이벤트로 나누기 vs interface 에서 함수로 분리하기 두가지중에서 개인적으로 제가 구현을 하는 것이었으면, 함수 두개로 분리하기를 택했을 수도 있을 것 같은데, 두개의 enum 으로 분리를 하셨던 이유가 있을까요?
두개를 비교해서, 이런점에서 enum 으로 분리를 하는 것이 좋았다!의 의견을 자유롭게 남겨주세요~💪

```kotlin
enum class ButtonEvent {
    INCREASE,
    DECREASE,
}
```

코드를 다시 짜게 된다면 함수 두 개로 분리하기를 선택할 거 같아요.🥲 처음에는 `increase()`, `decrease()` 함수로 구분할 필요 없이`update(event: ButtonEvent)`로 합쳐도 될 만큼 내부 로직이 커지지 않을 거라 생각했습니다.

시간이 지나니 단순히 수를 증감시키는 로직에도 '코루틴 스코프 열기', 'LiveData postValue', 'Repository에 CRUD' 등 로직이 복잡해지니, 차라리 증가-감소 함수를 분리하는 게 좋았을 거라는 생각이 들었어요.

 <!--
 리뷰어에게 질문이 있다면, PR에 직접 코멘트를 남겨 보세요.  
 특정 코드 줄에 의견을 남기면 보다 구체적인 피드백을 받을 수 있습니다.  
 -->

📌 [GitHub에서 PR에 댓글 남기는 방법](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request)

## 참고 자료

 <!-- 관련 문서나 참고할 만한 자료를 추가하세요. -->

## 스크린샷

 <!-- 필요하다면 스크린샷을 첨부하세요. -->

[Screen_recording_20250609_170945.webm](https://github.com/user-attachments/assets/8b2f2670-a8c3-4fea-be04-f979ae4eca2d)

ijh1298 committed:
refactor: ID를 Int -> Long 자료형으로 수정

ijh1298 committed:
refactor: CatalogRemoteDataSource 추상화 및 Repository와 분리

ijh1298 committed:
refactor: CartRemoteDataSource 추상화 및 Repository와 분리

ijh1298 committed:
refactor: 카탈로그 Repository, DataSource 코루틴 적용

ijh1298 committed:
refactor: 장바구니 Repository, DataSource, CatalogViewModel 코루틴 적용

ijh1298 committed:
refactor: 최근 본 상품 관련 Repository, ViewModel 비동기 처리 콜백 함수에서 코루틴으로 변경

- RecentlyViewedProductDao의 함수들을 suspend 함수로 변경
- RecentlyViewedProductRepository의 함수들을 suspend 함수로 변경
- DetailViewModel의 최근 본 상품 저장, 불러오기 로직을 코루틴으로 변경
- CartViewModel의 추천 상품 불러오기 로직을 코루틴으로 변경
- CatalogViewModel의 최근 본 상품 불러오기 로직을 코루틴으로 변경

ijh1298 committed:
test: JUnit5 CoroutinesTestExtension 추가

ijh1298 committed:
build: MockK 의존성 추가

ijh1298 committed:
test: CartViewModelTest 코루틴 테스트로 변경 및 Mock 객체 사용

ijh1298 committed:
test: CatalogViewModelTest 코루틴 테스트로 변경 및 Mock 객체 사용

ijh1298 committed:
test: DetailViewModelTest 코루틴 테스트로 변경 및 Mock 객체 사용

ijh1298 committed:
fix: CartItemService suspend 함수로 변경 및 RemoteCartProductRepositoryImpl 누락된 함수 추가

ijh1298 committed:
feat: 주문하기 클릭 시 장바구니 추천상품 Fragment로 전환하는 기능 구현

ijh1298 committed:
fix: 장바구니 아이템 삽입 후 CartItemId 반환토록 수정

ijh1298 committed:
feat: 장바구니 상품 수량 1개에서 감소 시 삭제 기능 추가

ijh1298 committed:
feat: 추천 Fragment 장바구니에 이미 추가된 상품이라면 미노출하도록 구현

ijh1298 committed:
chore: 사용하지 않는 코드 제거

ijh1298 committed:
feat: 주문 관련 DTO 클래스 추가

ijh1298 committed:
feat: 주문에 필요한 Coupon, AvailableTime Domain 모델 추가

ijh1298 committed:
feat: OrderActivity 및 레이아웃 추가

ijh1298 committed:
feat: CouponService, CouponDataSource, CouponRepository 구현

ijh1298 committed:
feat: 쿠폰 RecyclerView 적용

ijh1298 committed:
feat: OrderingProducts 일급 컬렉션 추가 및 주문 화면에 상품 목록 전달 기능 추가

ijh1298 committed:
feat: sealed class CouponType 추가

ijh1298 committed:
feat: OrderViewModel 생성자에 상품 전달하도록 구현

ijh1298 committed:
feat: OrderingProducts 모델 할인 금액 계산 로직 구현 및 레이아웃 ViewModel 연결

ijh1298 committed:
feat: 쿠폰 적용 기능 구현

- 쿠폰 선택 시 할인 금액이 변경되도록 구현
- 쿠폰을 다시 선택하면 선택이 해제되고 할인 금액이 원복되도록 구현
- 정률 할인 쿠폰의 할인 금액이 정상적으로 계산되도록 수정

ijh1298 committed:
refactor: 장바구니에서 주문 시 체크된 상품만 전달하도록 수정

ijh1298 committed:
feat: 쿠폰 선택 시 체크 표시 기능 추가 및 단일 선택되도록 수정

ijh1298 committed:
docs: README 기능 요구 사항 업데이트

ijh1298 committed:
refactor: 확장 함수 `parcelableArray`를 `IntentCompat.kt`로 이동

ijh1298 committed:
feat: 주문 기능 구현 및 주문 완료 시 상품 목록으로 이동

- OrderActivity
    - 주문 완료 시 상품 목록(CatalogActivity)으로 이동하고 기존 스택을 모두 제거하도록 `startActivityWhenOrderMade()` 메서드 추가
    - 주문하기 버튼 클릭 시 `OrderViewModel.makeOrder()` 호출
- OrderViewModel
    - `makeOrder()` 메서드 추가: `OrderRepository.insertOrders()`를 호출하여 주문을 시도하고, 결과를 `_isOrderMade` LiveData에 반영
- OrderRepository, OrderRepositoryImpl 추가
    - 주문 생성을 위한 `insertOrders()` 메서드 정의 및 구현
- OrderService 추가
    - 주문 생성을 위한 `/orders` POST API 정의
- Orders DTO 추가
    - 주문 생성 요청 시 cartItemIds를 담기 위한 DTO
- DataSourceProvider, RepositoryProvider
    - OrderService, OrderRepository 의존성 주입 설정
- OrderViewModelFactory
    - OrderRepository 의존성 주입 설정
- ProductAdapter
    - 불필요한 로그 제거

ijh1298 committed:
refactor: ProductAdapter에 DiffUtil 적용 및 ListAdapter로 변경

ijh1298 committed:
feat: 주문 상품에 쿠폰 만료일, 사용 가능 시간 적용

ijh1298 committed:
docs: README 요구 사항 업데이트

ijh1298 committed:
refactor: DependencyProvider를 DatabaseProvider로 이름 변경

ijh1298 committed:
refactor: CartActivity에서 프래그먼트 교체 로직을 메서드로 분리

ijh1298 committed:
refactor: CartSelectionFragment에서 `loadSelectAllState()` 호출 제거

불필요한 `loadSelectAllState()` 호출을 제거하여 코드 정리

ijh1298 committed:
fix: 장바구니 상품 체크 여부 반영

ijh1298 committed:
feat: 주문할 상품이 없을 경우 주문 버튼 비활성화

Gyuil-Hwnag commented on app/build.gradle.kts:
```diff
@@ -89,6 +89,7 @@ dependencies {
     implementation(libs.logging.interceptor)
     implementation(libs.androidx.fragment.ktx)
     implementation("org.jetbrains.kotlinx:kotlinx-coroutines-android:1.3.9")
+    testImplementation("io.mockk:mockk:1.14.2")
```
코멘트를 남기려고 하였지만, 이미 코드레빗이 코멘트를 남겨주었군요😅
[코드레빗 코멘트](https://github.com/woowacourse/android-shopping-order/pull/140/files#r2135242057)

ijh1298 replied:
version catalog에(`libs.versions.toml`) alias로 등록하는 것이 버전을 한 곳에 모아 관리하기 쉬워지겠네요!
여러 모듈에서 같은 라이브러리를 사용할 경우, 버전 충돌 가능성을 줄일 수 있다는 이점이 있었네요.

두루가 저번 피드백에서 남긴 Groovy 기반 의존성 관리와 현재의 version catalog 기반 의존성 관리에 대해
알아보라는 말을 이제 조금 이해한 것 같습니다!

[`build: kotlinx-coroutines-android, mockk 버전 라이브러리 카탈로그로 관리`](https://github.com/woowacourse/android-shopping-order/pull/140/commits/51346aa0581493d1a71a94a2703c4fe15f6bdeb1)

Gyuil-Hwnag replied:
맞아요~
아주 좋네요ㅎㅎ
이렇게 하나하나 알아가다보면 정말 많은 지식들이 쌓이게 되겠군요

Gyuil-Hwnag commented on app/src/main/AndroidManifest.xml:
```diff
@@ -5,18 +5,21 @@
         android:supportsRtl="true"
         android:theme="@style/Theme.Shopping"
+        android:usesCleartextTraffic="true"
         tools:targetApi="31">
```
이전에 넘어갔었나 보군요~
networkSecurityConfig 와 usesCleartextTraffic 는 각각 어떤 동작들을 할까요~?

ijh1298 replied:
> **networkSecurityConfig**
안드로이드 9(API 레벨 28)부터는 기본적으로 평문 트래픽인 HTTP가 차단되고, HTTPS를 통한 암호화된 트래픽 사용이 권장된다. 이러한 보안 강화는 사용자의 개인정보와 데이터를 보호하기 위해 필수적이다. 하지만 개발 및 테스트 단계에서는 HTTP를 사용할 필요가 종종 발생한다. 이럴 때 network_security_config를 사용하면 특정 도메인에 대해 HTTP를 예외적으로 허용할 수 있어 매우 유용하다.

> **usesCleartextTraffic**
앱이 일반 텍스트 HTTP와 같은 일반 텍스트 네트워크 트래픽을 사용하는지 여부를 나타냅니다. API 수준 27 이하를 타겟팅하는 앱에서는 기본값이 "true"이고, API 수준 28 이상을 타겟팅하는 앱에서는 기본값이 "false"입니다.
이 속성이 "false"로 설정되면 플랫폼 구성요소(예: HTTP 및 FTP 스택, DownloadManager, MediaPlayer는 앱의 일반 텍스트 트래픽 사용 요청을 거부합니다.

networkSecurityConfig가 HTTP를 허용하기 위한 도메인을 추가하는 작업이고, 
usesCleartextTraffic을 통해 HTTP 트래픽을 사용을 허용하는 작업임을 알았어요!

networkSecurityConfig에서는 도메인 하나하나에 대한 허용 여부를 설정할 수 있는데,
usesCleartextTraffic은 전역 허용 설정이었군요..! 개별 설정으로 수정하겠습니다.

[`refactor: localhost 도메인 제거 및 usesCleartextTraffic 속성 제거`](https://github.com/woowacourse/android-shopping-order/pull/140/commits/2ce2b294e26bb7285efa084083024fcd74bb0823)

Gyuil-Hwnag replied:
같이 있다보니.. 뭐지? 싶었어요ㅎㅎ
이번 기회를 통해서 해당 키워드에 대해서 어떤 동작을 하는지 알아갔으니, 다음번에는 좀 더 유용하게 사용을 해볼 수 있겠어요

Gyuil-Hwnag commented on app/src/main/AndroidManifest.xml:
```diff
@@ -5,18 +5,21 @@
         android:theme="@style/Theme.Shopping"
+        android:usesCleartextTraffic="true"
         tools:targetApi="31">
+        <activity
+            android:name=".order.OrderActivity"
+            android:exported="false" />
```
exported 속성은 어떤 동작을 하게될까요~?🤔

ijh1298 replied:
+ true인 경우, 모든 앱이 Activity에 접근할 수 있으며, 정확 클래스 이름으로 실행할 수 있다.
+ false인 경우, 동일한 애플리케이션, 동일한 사용자 ID를 가진 애플리케이션, 권한이 있는 시스템 구성요소에 의해서만 접근이 가능하다.

공식 문서에는 다음과 같이 적혀있습니다! LAUNCHER에 해당하는 Activity인 경우 시스템에 의해 
앱이 실행되어야 하므로 "true"를 주어야 한다고 알고 있어요. Activity를 생성하고 기본적으로 "false" 값이 적용되었습니다!

Gyuil-Hwnag replied:
좋습니다!
Activity Manifest 에는 여러가지 속성들이 있어요

이건 최근에 겪었던 이슈가 있어서 한번 참고차 남겨볼게요~
딥링크등으로 이동을 시키고, 해당`android:parentActivityName=""` 을 재대로 세팅을 안하다보니 딥링크로 Activity 실행시 순서가 보장이 안됐던 이슈가 있어서 한번 개념을 잡고 가면 좋을 것 같아서 남겨두었었습니다😀

Gyuil-Hwnag commented on README.md:
```diff
@@ -9,10 +9,18 @@
-- [ ] 해당 카테고리 상품이 10개 미만이라면 해당하는 개수만큼만 노출
-- [ ] 장바구니에 이미 추가된 상품이라면 미노출
-- [ ] 추천된 상품을 해당 화면에서 바로 추가하여 같이 주문할 수 있다.
+- [x] 장바구니 화면에서 특정 상품만 골라 주문하기 버튼을 누를 수 있다.
+- [x] 별도의 화면에서 상품 추천 알고리즘으로 사용자에게 적절한 상품을 추천해준다. (쿠팡 UX 참고)
+- [x] 상품 추천 알고리즘은 최근 본 상품 카테고리를 기반으로 최대 10개 노출한다.
```
https://github.com/user-attachments/assets/a441a270-0edf-4411-8607-aaa6310343cb

어라...? 제가 뭔가 잘못 실행을 시킨것일까요...?
저는 갑자기 이렇게 동작을 안하는군요..!
(에러 로그는 아래와 같습니다🥲)
<img width="1193" alt="스크린샷 2025-06-09 오후 9 36 46" src="https://github.com/user-attachments/assets/e0608f47-1c3e-4f65-a579-81d714d78664" />

ijh1298 replied:
최근 본 상품이 없을 때 최근 본 상품을 로드할 경우 발생하는 버그였군요..!! 감사합니다.😅

다음의 Room 쿼리를 하는 메서드에서, 
최근 본 상품이 DB에 없어서 null을 반환하는데 Long 타입으로 바꾸려해서 생기는 예외인 것 같습니다..!

```kotlin
@Query("SELECT productUid FROM RecentlyViewedProducts ORDER BY timestamp DESC LIMIT 1")
suspend fun getLatestViewedProductId(): Long?
```

수정했습니다!

[`fix: 최근 본 상품이 없을 때, 최근 본 상품을 로드할 경우 발생하는 버그 수정`](https://github.com/woowacourse/android-shopping-order/pull/140/commits/03eb663e921823c8a8e022f736395babf30f13d3)

Gyuil-Hwnag commented on app/src/main/java/woowacourse/shopping/cart/CartActivity.kt:
```diff
@@ -12,14 +11,15 @@ import androidx.core.view.ViewCompat
     val viewModel: CartViewModel by viewModels { CartViewModelFactory() }
     private var hasHandledTotalCount = false
+    private var fragmentState: Int = FRAGMENT_SELECTION
 
```
이 부분이 마음에 걸리기는 하는군요~
flag 값으로 지정하는 방법외에 다른 방법들도 있을까? 라는 생각이 들었습니다🤔
fragment를 비교해서 가져오는 방법들도 있을 것 같아서요!

ijh1298 replied:
FragmentManager를 통해 현재 Attach된 Fragment를 알아내는 방식도 있었네요..💡두루 덕분에 새로운 방식을 알아갑니다!!
제가 선택한 FLAG 방식을 사용하다 지정해주는 걸 놓치는 실수를 방지할 수 있을 듯 합니다!

```kotlin
  private fun processOrderClick(orderedProducts: Array<ProductUiModel>) {
      val currentFragment =
          supportFragmentManager.findFragmentById(R.id.fragment_container_cart_selection)

      when (currentFragment) {
          is CartSelectionFragment -> {
              replaceCartRecommendationFragment()
          }

          is CartRecommendationFragment -> {
              startActivity(OrderActivity.newIntent(this, orderedProducts))
          }
      }
  }
```

[`refactor: 주문하기 버튼 클릭 시 프래그먼트 상태 비교 로직 수정`](https://github.com/woowacourse/android-shopping-order/pull/140/commits/ba9517651a6d0c872d2a4554ccc320e7fa839a58)

Gyuil-Hwnag replied:
오호 아주 좋습니다~👍
원했던 내용을 너무 잘 찾아주셨군요!

Gyuil-Hwnag commented on app/src/main/java/woowacourse/shopping/cart/CartViewModel.kt:
```diff
@@ -45,20 +55,33 @@ class CartViewModel(
     val updatedProduct: LiveData<ProductUiModel> = _updatedProduct
 
+    private val _orderClicked = MutableLiveData<Unit>()
+    val orderClicked: LiveData<Unit> = _orderClicked
```
orderedProducts 해당 변수와, orderClicked 해당 변수를 두개로 나누는 것이 아닌 한개로 묶을 수도 있었겠다는 생각이 드는데 크림은 어떠신가요~?

ijh1298 replied:
충분히 묶어서 활용할 수 있었군요..! 좋은 방식 알려주셔서 감사합니다.👍
[`refactor: CartViewModel.orderClicked LiveData의 값으로 주문할 상품 목록을 전달`](https://github.com/woowacourse/android-shopping-order/pull/140/commits/3761363ccd5fb9cc714752cb12f4f5926f03342a)

Gyuil-Hwnag commented on app/src/main/java/woowacourse/shopping/cart/CartViewModel.kt:
```diff
@@ -45,20 +55,33 @@ class CartViewModel(
+        viewModelScope.launch {
+            val latestViewedProduct: ProductUiModel? =
+                recentlyViewedProductRepository.getLatestViewedProduct()
+            latestViewedProduct?.id ?: return@launch
```
만약 Repository 함수들이 에러가 발생하게 된다면, 사용자는 어떠한 안내도 없이 그저 동작을 안되는 것처럼 보이는 것일까요~🤔

Gyuil-Hwnag replied:
추가로 위에서 제가 실행을 하면서, 남겨두었던 에러 로그를 보면 Repository 의 함수에서 크래시가 발생을 하는 경우 어떻게 캐치를 하게 할 수 있을까요?!

ijh1298 replied:
기존 코드가 우선 이상하게 작성되어 있다는 것을 인지했습니다.🥲
Repository에서 예외가 발생하는 경우 try-catch 문으로 감싸기만 했을 뿐, 에러에 대해 사용자가 아무런 응답을 받지 못하고 있었어요.

단지 `emptyList()` 같은 것을 전달해서, 앱이 크래시나지만 않도록 하고 있었습니다. (두루가 발견해주신 예외는 크래시가 났군요..)

따라서 `runCatching`을 repository 메서드들에 적용해줌으로써, 
예외가 발생하는 경우 failure 값이 true인 `Result`를 반환하도록 했습니다.

viewModel에서 **Result가 false인 경우**에 viewModelScope를 메서드 실행을 중단하도록 (도중에 return) 수정하고,
사용자가 예외가 발생해서 정상 로드가 되지 않는다는 것을 인지할 수 있도록 토스트 메시지를 띄우도록 수정했습니다!

[`refactor: 네트워크 통신 결과 Result 타입으로 변경 및 에러 처리 추가`](https://github.com/woowacourse/android-shopping-order/pull/140/commits/44eae2971c7646b4c7f1582ce67eff0fa6bf1aac)
[`refactor: Retrofit Service Response 반환값 제거 및 CouponRepository, OrderRepository에 Result 타입으로 변경 및 에러 처리 추가`](https://github.com/woowacourse/android-shopping-order/pull/140/commits/44ec1b03a562bb72c33e4681b7bf2f2b423a6338)

Gyuil-Hwnag commented on app/src/main/java/woowacourse/shopping/data/datasource/CartRemoteDataSourceImpl.kt:
```diff
@@ -0,0 +1,101 @@
+            } else {
+                throw Exception("서버 응답 실패 : ${response.code()}")
+            }
+        }
```
`withContext(Dispatchers.IO)`로 작성을 하신 이유가 있을까요🤔

Gyuil-Hwnag replied:
추가로 상위에서 Exception이 발생할 경우 최종적으로 함수의 결과는 null 이 될텐데, 이 방법으로 적용을 하게 된다면 모든 함수의 결과에서는 null check 를 통해서 에러에 대한 체크를 해야 하는데, 위에 Repository 에 남겨둔 내용 처럼 처리를 하는 방법은 어떤 것이 있을까요?!

ijh1298 replied:
예외 처리를 수정한 내역은 위 코멘트에 달아두었습니다..!

> withContext(Dispatchers.IO)로 작성을 하신 이유가 있을까요🤔

잘 몰라서 작성한 듯합니다..

Room, Retrofit 라이브러리는 자체적으로 비동기 처리를 하기 때문에 `Dispatchers.IO`를 굳이 하지 않아도 된다고 배웠어요.
불필요하게 withContext를 통해 컨텍스트를 전환하면 오버헤드가 발생할 수도 있겠군요, 수정했습니다!

[`refactor: DataSource 코드 withContext 제거`](https://github.com/woowacourse/android-shopping-order/pull/140/commits/ffa5df2c341f3bf57143e29dcb71b2c215b95110)

Gyuil-Hwnag replied:
Coroutine 으로 넘어오면서 내부적으로 Thread 를 스위칭 해주는 동작들이 있어요~
이에 대한 원리는 한번 추후에 살펴본다면, 크림의 강점으로도 가져가 볼 수 있겠군요!

Gyuil-Hwnag commented on app/src/main/java/woowacourse/shopping/data/dto/order/Coupons.kt:
```diff
@@ -0,0 +1,3 @@
+package woowacourse.shopping.data.dto.order
+
+class Coupons : ArrayList<Coupon>()
```
오호~ 코드레빗이 이런 내용도 코멘트로 남길 수가 있군요
한번 살펴보면 괜찮은 내용인 것 같습니다!
[코드레빗 코멘트](https://github.com/woowacourse/android-shopping-order/pull/140/files#r2135242067)

저는 여기서 추가적으로 ArrayList 로 설정을 해야 했을까? 라는 생각이 들었어요

ijh1298 replied:
안드로이드 환경에서 DTO를 처음 만들어봐서, API 명세에 따라 객체를 만들었습니다.😓
`List<Coupon>`으로 대체가 되는 군요.. 감사합니다!! 제거했습니다.

[`refactor: Coupons를 List<Coupon>으로 변경`](https://github.com/woowacourse/android-shopping-order/pull/140/commits/0b981c6df4ebbf5466a668772c44077ac65d8cfc)

Gyuil-Hwnag commented on app/src/main/java/woowacourse/shopping/data/repository/CouponRepositoryImpl.kt:
```diff
@@ -0,0 +1,16 @@
+        } catch (e: Exception) {
+            emptyList()
+        }
+}
```
이전에 남겨둔 코멘트들에 대한 힌트가 될 것 같아서, 참고해보면 좋을 것 같다는 생각이 듭니다~💪
[코드 레빗 코멘트](https://github.com/woowacourse/android-shopping-order/pull/140/files#r2135242084)

Result 를 사용하는 방식외에 CoroutineExceptionHandler 를 활용해보는 방법이 있을 것 같습니다!

ijh1298 replied:
우선은 Result를 사용하는 방식이 더 적용하기 쉬워보여 `Result`를 사용하여 리팩터링을 진행해보았습니다!

Gyuil-Hwnag replied:
좋습니다~ CoroutineExceptionHandler 로 하는 방법도 있으니, 추후에 시간이 난다면 한번 살펴보는 것도 좋을 것 같습니다💪

Gyuil-Hwnag commented on app/src/main/java/woowacourse/shopping/data/repository/OrderRepositoryImpl.kt:
```diff
@@ -0,0 +1,19 @@
+            } else {
+                throw HttpException(response)
+            }
+        }
```
이 부분에 대해서 코드레빗도 코멘트를 남겼군요~
[코드 레빗 코멘트](https://github.com/woowacourse/android-shopping-order/pull/140/files#r2135242093)

`reponse.body()`와 같은 형태로 작성을 하게 되신 이유가 궁금합니다!

ijh1298 replied:
Kotlin 기본 지식이 부족했네요..🥲
리팩터링을 진행하면서 Repository에서 예외가 발생할 수 있는 부분을 모두 `runCatching`으로 감싸는 작업을 진행했습니다.

그러면서 runCatching의 반환값이 `Result`라는 것을 알게 되었습니다. 따라서 내부에서 성공/실패 여부를 구분하여 throw 할 필요가 없었네요! `orderService`에서 예외가 발생할 경우 Result가 Failure로 전달될 것이었군요. `response.body()`가 반환값이 없는 API를 호출하고 있어서, Result 안에 그것을 넣어주는 형태로 사용하게 되었는데 의미 없는 행위인 듯합니다..

코드를 수정했습니다!

[`refactor: 네트워크 통신 결과 Result 타입으로 변경 및 에러 처리 추가`](https://github.com/woowacourse/android-shopping-order/pull/140/commits/44eae2971c7646b4c7f1582ce67eff0fa6bf1aac)
[`refactor: Retrofit Service Response 반환값 제거 및 CouponRepository, OrderRepository에 Result 타입으로 변경 및 에러 처리 추가`](https://github.com/woowacourse/android-shopping-order/pull/140/commits/44ec1b03a562bb72c33e4681b7bf2f2b423a6338)

Gyuil-Hwnag commented on app/src/main/java/woowacourse/shopping/data/service/CartItemService.kt:
```diff
@@ -14,28 +14,28 @@ import woowacourse.shopping.data.dto.cartitem.UpdateCartItemRequest
-        @Query("page") page: Int = 0,
-        @Query("size") size: Int = 1,
+    suspend fun requestCartItems(
+        @Query("page") page: Int? = 0,
+        @Query("size") size: Int? = 1,
         @Query("sort") sort: List<String> = listOf(),
-    ): Call<ProductResponse>
+    ): Response<ProductResponse>
```
`Response<ProductResponse>`와 `ProductResponse`로 리턴을 받는 방법들이 있었을텐데, Response<T> 형태로 반환값을 지정하게 되신 이유가 있을까요~

ijh1298 replied:
코루틴을 사용하기 이전에는 반환값을 Call로 해주었습니다. 
백그라운드 스레드의 비동기로 요청하기 위해 `enque` 메서드를 사용하기 위함이었습니다.

코루틴을 사용하면서 Call 대신 `Reponse`를 사용할 수 있음을 알았습니다! (Response 없이 DTO로 받을 수 있는지 몰랐네요..)
`Response` 타입으로 감싼 반환값을 사용하면, `body()`, `code()`, `isSuccessful` 같은 자세한 값을 확인할 수 있었습니다.

리팩터링을 진행하면서 구체적으로 에러 코드를 처리하는 분기를 만들 계획은 아직 없었기에,
`ProductResponse`를 반환하게 하고, 에러가 발생할 수 있는 경우는 `Repository`의 `runCatching`에서 잡도록 수정했습니다!

다만 `Response`에서 `Header`의 `location`을 따와 cartItemId를 추출하는 등 Response가 필요한 로직도 있었는데요, (`CartItemService`의 `postCartItem` 메서드 등) 그 부분은 반환값을 Reponse로 유지했습니다!

[`refactor: Retrofit Service Response 반환값 제거 및 CouponRepository, OrderRepository에 Result 타입으로 변경 및 에러 처리 추가`](https://github.com/woowacourse/android-shopping-order/pull/140/commits/44ec1b03a562bb72c33e4681b7bf2f2b423a6338)

Gyuil-Hwnag replied:
사실 이부분은 Coroutine 으로 넘어오면서, ProductResponse 로 가능해질 수 있겠어요~
추가적으로 Result<T> or CoroutineExceptionHandler 를 사용한다면, Call 로 받았을 때의 에러에 대한 대응도 가능해질 것 같습니다

Gyuil-Hwnag commented on app/src/main/java/woowacourse/shopping/di/DataSourceProvider.kt:
```diff
@@ -0,0 +1,34 @@
+        RetrofitProductService.INSTANCE.create(CartItemService::class.java)
+
+    private val couponServcie: CouponService =
+        RetrofitProductService.INSTANCE.create(CouponService::class.java)
```
오타가 있었나 보군요🥲
[코드 레빗 코멘트](https://github.com/woowacourse/android-shopping-order/pull/140/files#r2135242106)

ijh1298 replied:
수정했습니다!🥲
[fix: CouponService 변수명 오타 수정](https://github.com/woowacourse/android-shopping-order/pull/140/commits/9c69c344b990df313edf70a92c51686409b30233)

Gyuil-Hwnag commented on app/src/main/java/woowacourse/shopping/order/OrderViewModel.kt:
```diff
@@ -0,0 +1,91 @@
+                _isOrderMade.postValue(true)
+            } else {
+                _isOrderMade.postValue(false)
+            }
```
postValue 와 setValue는 어떠한 차이점이 있을까요?🤔

ijh1298 replied:
`setValue()`는 메인 스레드에서 동작하며 LiveData 값을 즉시 변경합니다.
`postValue()`는 백그라운드 스레드에서 동작하며 LiveData 값을 비동기적으로 변경합니다.

`viewModelScope`가 기본적으로 메인(UI) 스레드에서 동작하니,
 저 코드는 굳이 `postValue()`를 사용하지 않아도 됐네요..! 수정했습니다!

[`refactor: viewModelScope 내 LiveData postValue 대신 setValue 사용`](https://github.com/woowacourse/android-shopping-order/pull/140/commits/207f6912347b66fc18b1b33ef4f073d690123e64)

Gyuil-Hwnag reviewed (changes requested):
크림 안녕하세요~
해당 미션 구현 고생 많으셨습니다!

코루틴으로 넘어오면서, 많이 어려우셨을텐데 적용 정말 고생 많으셨습니다~
마지막 미션인만큼 모든 부분에 대한 코멘트를 남겨두는 것 보다는 같이 코루틴에 대한 얘기를 해보면 좋을 것 같다는 생각이 들었습니다.
해당 부분들에 대해서 코멘트를 남겨두었으니, 확인해주시면 감사하겠습니다!

궁금한 점이나 논의하고 싶은 내용이 있으면 언제든 DM 주세요! 🙏
마지막까지 화이팅입니다💪

ijh1298 committed:
build: kotlinx-coroutines-android, mockk 버전 라이브러리 카탈로그로 관리

ijh1298 committed:
refactor: localhost 도메인 제거 및 usesCleartextTraffic 속성 제거

ijh1298 committed:
refactor: viewModelScope 내 LiveData postValue 대신 setValue 사용

ijh1298 committed:
fix: CouponService 변수명 오타 수정

ijh1298 committed:
refactor: 네트워크 통신 결과 Result 타입으로 변경 및 에러 처리 추가

- Repository 및 ViewModel에서 네트워크 통신 결과를 Result 타입으로 변경했습니다.
- 통신 실패 시 사용자에게 Toast 메시지를 통해 에러를 알리는 기능을 추가했습니다.

ijh1298 committed:
refactor: Retrofit Service Response 반환값 제거 및 CouponRepository, OrderRepository에 Result 타입으로 변경 및 에러 처리 추가

ijh1298 committed:
refactor: DataSource 코드 withContext 제거

Retrofit의 Call Adapter는 이미 IO 스레드에서 네트워크 요청을 처리하므로 withContext(Dispatchers.IO)를 제거합니다.

ijh1298 committed:
fix: 최근 본 상품이 없을 때, 최근 본 상품을 로드할 경우 발생하는 버그 수정

ijh1298 committed:
refactor: Coupons를 List<Coupon>으로 변경

ijh1298 committed:
refactor: CartViewModel.orderClicked LiveData의 값으로 주문할 상품 목록을 전달

ijh1298 committed:
refactor: 주문하기 버튼 클릭 시 프래그먼트 상태 비교 로직 수정

- 기존의 fragmentState 변수를 사용한 상태 비교에서, 현재 표시되고 있는 프래그먼트의 타입을 직접 확인하는 방식으로 변경했습니다.
- 이는 프래그먼트 상태 관리의 명확성을 높이고 잠재적인 오류를 줄이기 위함입니다.

ijh1298 reviewed (commented):
좋은 피드백 남겨주셔서 정말 감사합니다, 두루! 🙇🏻
아직 개선할 게 산더미지만, 피드백을 우선시 반영해보았습니다!

Gyuil-Hwnag commented on app/src/main/java/woowacourse/shopping/cart/CartViewModel.kt:
```diff
@@ -78,38 +109,68 @@ class CartViewModel(
     fun addProduct(product: ProductUiModel) {
-        cartProductRepository.insertCartProduct(product.copy(quantity = 1)) { product ->
-            _updatedProduct.postValue(product)
+        viewModelScope.launch {
+            val cartItemId =
+                cartProductRepository.insertCartProduct(productId = product.id, quantity = 1).getOrElse {
+                    _errorMessage.setValue("상품 추가에 실패했습니다.")
+                    return@launch
+                }
+            val addedProduct = product.copy(quantity = 1, cartItemId = cartItemId)
+            _updatedProduct.value = addedProduct
```
이 부분은 코멘트로 남겨두었던 대로 CoroutineExceptionHandler 를 활용하는 방법도 있으니, 한번 살펴보셔도 좋을 것 같습니다!
~~~kotlin
  fun addProduct(product: ProductUiModel) {
        // 기존 방식
        viewModelScope.launch {
            val cartItemId =
                cartProductRepository.insertCartProduct(productId = product.id, quantity = 1).getOrElse {
                    _errorMessage.setValue("상품 추가에 실패했습니다.")
                    return@launch
                }
            val addedProduct = product.copy(quantity = 1, cartItemId = cartItemId)
            _updatedProduct.value = addedProduct
            refreshProductsInfo()
            loadCartProducts()
        }
        
        // ExceptionHandler 방식
        viewModelScope.launch(CoroutineExceptionHandler { _, e -> 
            _errorMessage.setValue("상품 추가에 실패했습니다: ${e.message}")
        }) { 
            val cartItemId = cartProductRepository.insertCartProduct(productId = product.id, quantity = 1)
            val addedProduct = product.copy(quantity = 1, cartItemId = cartItemId)
            _updatedProduct.value = addedProduct
            refreshProductsInfo()
            loadCartProducts()
        }
    }
~~~

Gyuil-Hwnag commented on app/src/main/java/woowacourse/shopping/data/datasource/CartRemoteDataSourceImpl.kt:
```diff
@@ -0,0 +1,77 @@
+        return if (response.isSuccessful) {
+            val locationHeader = response.headers()["location"]
+            val id = locationHeader?.substringAfterLast("/")?.toLongOrNull()
+            id ?: throw Exception("cartItemId 가져오기에 실패했습니다.")
+        } else {
+            throw Exception("서버 응답 실패 : ${response.code()}")
+        }
```
Exception 들을 sealed class를 통해서 아래와 같은 방식으로 사용을 한다면, 특정 서버 에러에 대해서는 판별해서 분기처리등으로도 활용을 하여 로직을 개선해볼 수 있겠군요!
~~~kotlin
class LoadCardException(val message: String): Exception()
~~~

Gyuil-Hwnag commented on app/src/main/java/woowacourse/shopping/data/dto/cartitem/CartItemContent.kt:
```diff
@@ -0,0 +1,23 @@
+    val quantity: Int,
+)
+
+fun CartItemContent.toUiModel(): ProductUiModel =
+    ProductUiModel(
+        id = product.id,
+        imageUrl = product.imageUrl,
+        name = product.name,
+        price = product.price,
+        cartItemId = id,
+        quantity = quantity,
+    )
```
Mapper 부분들의 생성 정말 좋습니다~

Mapper 와 Domain 객체의 companion object 내에 생성을 하는 방법 두가지가 있었을 것 같은데, 각각은 어떤 장단점이 있을지 비교를 해보고 크림만의 기준을 잘 세워보는 것도 좋을 것 같습니다!

개인적으로는 companion object 내에 생성을 하는 편이기는 했어요

이유는 점점 프로젝트가 커지면서, 여러 Response를 해당 객체로 변환을 하게 되는데 이렇게 하는 것이 좀 더 가시성을 증가시켜주는 측면도 있었어요

Gyuil-Hwnag commented on app/src/main/java/woowacourse/shopping/order/OrderViewModel.kt:
```diff
@@ -0,0 +1,96 @@
+    private val _availableDisplayingCoupons = MutableLiveData<List<CouponUiModel>>(emptyList())
+    val availableDisplayingCoupons: LiveData<List<CouponUiModel>> = _availableDisplayingCoupons
+
+    private lateinit var availableCoupons: List<Coupon>
+
```
이 부분은 코드레빗의 코멘트를 한번 살펴봐도 좋을 것 같다는 생각이 듭니다~
[코드 레빗 코멘트](https://github.com/woowacourse/android-shopping-order/pull/140/files#r2135242149)]
개인적으로 lateinit var 은 지양을 하는 편이기는 했어요!
이유는 간단하게는 개발자의 실수등도 집어볼 수 가 있겠군요

Gyuil-Hwnag reviewed (approved):
크림 안녕하세요~
미션 너무 고생 많으셨습니다! 👏
리뷰 반영 정말 잘 해주셨어요💯

구현을 하시면서 여러가지 고민을 하셨을 텐데 다음번 미션 에서도 잘 적용을 하면서 발전해 나가시길 바라겠습니다😆

해당 미션은 이만 여기서 마무리하고 머지하도록 하겠습니다.
도움이 될만한 내용들 코멘트 달았으니 확인해보셔도 좋을 것 같습니다!

궁금한 점이나 논의하고 싶은 내용이 있으면 언제든 DM 주세요! 🙏
레벨1, 2기간 동안 정말 고생 많으셨습니다~
앞으로도 응원하겠습니다 💪
